src/Compilation.zig

```diff
@@ -1366,6 +1375,21 @@ pub fn getAllErrorsAlloc(self: *Compilation) !AllErrors {
     };
 }
 
+const WaitGroup = struct {
```

Comment:
There might be a race here @marler8997 

* WaitGroup.done() decrements counter to 0, gets preempted
* WaitGroup.wait() sees 0 counter, exits, deallocates itself if on stack
* WaitGroup.done() gets rescheduled, saw the 0 dec, tries to set event, event memory was already deallocated

Reply:
Feel like you'd need sort of sort of futex wait comparison/validation mechanic to resolve this. Wrapping all the state checking in a Mutex would be the cross platform equivalent. Might be a good reason to introduce [parking_lot](https://webkit.org/blog/6161/locking-in-webkit/) in the stdlib.

Reply:
Yeah I haven't reviewed all this code, but based on your comment it does look like a race.

P.S. It seems like reviewing multi-threaded scheduler code takes about 10 times longer than it does to review normal code.  It's going to take me some time to really review all this code, even though it's less than 200 lines!

Reply:
Yea, I plan to do a rewrite of this later today with the goal of simplifying the review process at the possible expense of concurrent perf.